libphal/phal_pdbg.C

```diff
@@ -300,24 +301,134 @@ bool isSbeVitalAttnActive(struct pdbg_target *proc)
+constexpr int reqAttrFound = 1;
+constexpr int reqAttrNotFound = 2;
+LocationCode g_locCode;
+
+std::optional<LocationCode> getUnexpandedLocCode(const std::string &locCode)
+{
+
+	// "-" should be there to seggregate all (FC, Node number and SE values)
+	// together.
+	// Note: Each (FC, Node number and SE) value can be seggregate by "."
+	// but, cec device tree just have unexpand format so, just skipping
+	// still first occurrence "-" and replacing with "fcs".
+	auto endPosOfFcs = locCode.find('-', EXP_LOCATIN_CODE_MIN_LENGTH);
+	if (endPosOfFcs == std::string::npos) {
+		log(level::ERROR,
+		    "Given location code %s is not valid i.e could not find "
+		    "dash",
+		    locCode);
+		return std::nullopt;
+	}
+
+	std::string unExpandedLocCode("Ufcs");
+
+	if (locCode.length() > EXP_LOCATIN_CODE_MIN_LENGTH) {
+		unExpandedLocCode.append(
+		    locCode.substr(endPosOfFcs, std::string::npos));
+	}
+
+	return unExpandedLocCode;
+}
+
+int pdbgCallbackToGetLocCode(struct pdbg_target *target, void *fruType)
+{
+	ATTR_LOCATION_CODE_Type locCode;
+	std::vector<uint8_t> frus{ENUM_ATTR_TYPE_DIMM, ENUM_ATTR_TYPE_PROC,
+				  ENUM_ATTR_TYPE_TPM, ENUM_ATTR_TYPE_BMC};
+	if (DT_GET_PROP(ATTR_LOCATION_CODE, target, locCode))
+		return continueTgtTraversal;
+	else {
+		if (locCode == g_locCode) {
+			if (!DT_GET_PROP(ATTR_TYPE, target,
+					 *static_cast<int *>(fruType)) &&
+			    std::find(frus.begin(), frus.end(),
+				      *static_cast<int *>(fruType)) !=
+				frus.end())
+				return reqAttrFound;
+			return reqAttrNotFound;
+		}
+		return continueTgtTraversal;
+	}
+	return reqAttrNotFound;
+}
+
+/**
+
+ * @brief Retrieves FRU information for the provided location codes.
+
+ *              Location codes can be in either expanded or unexpanded format.
+
+ *
+
+ * @param[in] locationCode A FRU location code to fetch its type.
+
+ *
+
+ * @return FRUType or error code
+
+ */
+
+std::expected<FRUType, openpower::phal::exception::ERR_TYPE>
+    getFRUType(const LocationCode &locationCode) noexcept
+{
+	// Location code should start with "U"
+	g_locCode = locationCode;
+	if (locationCode[0] != 'U') {
+		log(level::ERROR,
+		    "Location code should start with \"U\" but, given location "
+		    "code %s",
+		    locationCode);
+		return std::unexpected(exception::INVALID_LOC_CODE);
+	}
+	if (locationCode.length() >= EXP_LOCATIN_CODE_MIN_LENGTH) {
+
+		auto loc = getUnexpandedLocCode(locationCode);
+		if (loc.has_value())
+			g_locCode = loc.value();
+		else {
+			return std::unexpected(exception::INVALID_LOC_CODE);
+		}
+	}
+	if (locationCode.length() < UNEXP_LOCATIN_CODE_MIN_LENGTH)
+		return std::unexpected(exception::INVALID_LOC_CODE);
+	ATTR_TYPE_Type fruType = ENUM_ATTR_TYPE_NA;
+	auto ret = pdbg_target_traverse(
+	    nullptr /* Passing NULL to start target traversal from root */,
+	    pdbgCallbackToGetLocCode, &fruType);
+	if (ret == 1)
+		return fruType;
+	else
+		return std::unexpected(exception::LOC_CODE_NOT_FOUND);
```

Comment:
Please consider using the error code mentioned in LN378. Currently, we are unable to identify the API bug where "the user passed a valid location code, but it does not exist in the predefined FRUs list" or when the location code attribute does not exist for the predefined FRUs target.